util/VarText.cpp

```diff
@@ -93,8 +93,12 @@ namespace {
             return boost::none;
         }
         T* object = GetByID(id);
-        if (!object)
-            return boost::none;
+        if (!object) {
+            if (std::is_same<T, const ShipDesign>::value)
+                return UserString("FW_UNKNOWN_DESIGN_NAME");
+            else
+                return UserString("TT_UNKNOWN");
```

Comment:
> this seems to contradict the basic premise of this function, which was to return a boost::none if the requested design isn't available

Please explain how you extract that from the documentation
```
    /// Returns substitution string for a ship design tag
```
or from the actual use, or wherever you get it from.  It looks to me like it is simply a minor aspect of how someone happened to implement it.  And it still returns boost::none in some cases, but just not when there is a better string to return.

> presumably calling code should check for the boost::none and react appropriately.

The only 2 places where this code is currently used are right here:
https://github.com/freeorion/freeorion/blob/260abc9b90c2dde42578c04498557c2743777b5d/util/VarText.cpp#L149-L153

I think those uses are better off left more streamlined, dealing with this 'unknown' aspect here.

Reply:
I get it from looking at the code. If it was supposed to return a string in all cases, including "Unknown" or similar when the ID wasn't a ship design id, then it would just return a string in all cases, with appropriate default values in those cases. Instead, it return an optional, which is left as boost::none in case where the text can't be casted to an integer or when there is no such design. If you prefer it to return a string in cases where it fails to find a design for the specified text, it should do that always and not return an optional.

Reply:
> I get it from looking at the code.... 

so it's a "basic premise" just because that's how it has been working for a while?   But it would be fine to totally destroy it? (I'll confess I am particularly suspicious of my understanding of that last part.)  And we can't make a partial change to the "basic premise," not because the partial change would cause any problems anywhere, just... because?  Is that a rule I'm supposed to apply generally, or just in this particular case?

I propose that we just redefine the "basic premise" to be that we return boost::none only for the major error event that the passed ID string could not even be converted to an int, and we return suitable user friendly strings for lesser problems where we have a suitably friendly string.

Reply:
It's a "basic premise" because that's the function declaration. I don't mean that's how the function has to be implemented, but just that that's what it does now. I also don't mean "basic premise" as some specifically defined technical term for how function definitions and implementation work... I just needed a term to describe what the function does and its declared format and existing implementation.

It's fine to change it so it returns strings instead of optionals, and I'd have used placeholders in case of errors instead of optionals were I to have implemented this (or the "valid" parameter that was used before the optionasl). But I don't think it's a good idea to sometimes return `none` and sometimes return a placeholder string in case of an error or unexpected situation.